File: backend/app/core/blinding_type.py
from enum import IntEnum


class BlindingType(IntEnum):
    OPEN_LABEL = 0
    PB = 1  # Participant blinding
    PIB = 2  # Participant and Investigator blinding
    PISB = 3  # Participant, Investigator and Statistician blinding


BLINDING_TYPE_LABELS: dict[BlindingType, str] = {
    BlindingType.OPEN_LABEL: "Open Label",
    BlindingType.PB: "Participant blinding (PB)",
    BlindingType.PIB: "Participant and Investigator blinding (PIB)",
    BlindingType.PISB: "Participant, Investigator and Statistician blinding (PISB)",
}

# PISB is equivalent to PIB from the investigator's perspective.
INVESTIGATOR_BLINDED_BLINDING_TYPES = frozenset({BlindingType.PIB, BlindingType.PISB})


def validate_blinding_type(value: int) -> int:
    try:
        BlindingType(value)
    except ValueError:
        raise ValueError("blinding_type must be 0 (Open Label), 1 (PB), 2 (PIB), or 3 (PISB)")
    return value


def investigator_is_blinded(blinding_type: int) -> bool:
    return blinding_type in INVESTIGATOR_BLINDED_BLINDING_TYPES

File: frontend/src/utils/blindingType.js
export const BLINDING_TYPE = {
  OPEN_LABEL: 0,
  PB: 1,
  PIB: 2,
  PISB: 3,
}

export const DEFAULT_BLINDING_TYPE = BLINDING_TYPE.PIB

// PISB is equivalent to PIB from the investigator's perspective.
export const INVESTIGATOR_BLINDED_BLINDING_TYPES = [
  BLINDING_TYPE.PIB,
  BLINDING_TYPE.PISB,
]

export const BLINDING_TYPE_OPTIONS = [
  { value: BLINDING_TYPE.OPEN_LABEL, label: 'Open Label' },
  { value: BLINDING_TYPE.PB, label: 'Participant blinding (PB)' },
  { value: BLINDING_TYPE.PIB, label: 'Participant and Investigator blinding (PIB)' },
  {
    value: BLINDING_TYPE.PISB,
    label: 'Participant, Investigator and Statistician blinding (PISB)',
  },
]

const BLINDING_TYPE_SHORT_LABELS = {
  [BLINDING_TYPE.OPEN_LABEL]: 'Open Label',
  [BLINDING_TYPE.PB]: 'PB',
  [BLINDING_TYPE.PIB]: 'PIB',
  [BLINDING_TYPE.PISB]: 'PISB',
}

export function blindingTypeLabel(value) {
  const option = BLINDING_TYPE_OPTIONS.find((item) => item.value === value)
  return option?.label ?? `Unknown (${value})`
}

export function blindingTypeShortLabel(value) {
  return BLINDING_TYPE_SHORT_LABELS[value] ?? `Unknown (${value})`
}

export function investigatorIsBlinded(blindingType) {
  return INVESTIGATOR_BLINDED_BLINDING_TYPES.includes(blindingType)
}
